data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/attacks/attacks_zero.cfg

```diff
@@ -0,0 +1,31 @@
+# wmllint: no translatables
+
+#####
+# API(s) being tested: [attacks]value=
+##
+# Actions:
+# Give the leaders an attacks ability with a value of 0.
+# Have the side 1 leader attack the side 2 leader with both its weapons
+##
+# Expected end state:
+# The side 1 leader has 2 weapons both of which don't have 0 strikes
```

Comment:
they should have 0 strikes then?

Reply:
The `COMPARE=numerical_not_equals` means it's checking that the strikes aren't equal to 0.

Reply:
I think it is bug if unit has nonzero strikes after this ability.

Reply:
Feel free to open an issue for it then - the point of these tests is just to prove what the current state of things actually is.

Reply:
Is it possible to know which tests were written based on preserving important functionality and which just to test the status quo?

I think in the future it's very likely to think that certain behavior must be sane and wanted just because there is a test for it.

Reply:
I understand all unit tests as being about protection from unintentional changes, while not restricting anything about intentional changes.

Reply:
I think it would be good to comment in the "expected end state" if the end state is illogical, but it might be better to just log a bug about it.

Reply:
There isn't, and I think that for the most part it's a moot question - if you break compatibility then absent confirmation otherwise from searching the add-ons server you're probably breaking at least a few people's code, regardless of whether there's a unit test for it or not.

Reply:
My point was not about whether compatibility is broken. It is about whether it may be a good idea to break compatibility or not.

If every dev sees tests as nothing more than preventing unintentionally changing current behavior then it indeed does not matter. I doubt that is the case though but if that is stated officially somewhere that'd help.

---

IMO though when writing tests is a good time to think about whether whatever behavior is being tested is actually sane. Especially for AbilitiesWML with known abysmal documentation just mindlessly writing tests does not seem that useful since you can't tell whether the tests agree with the documentation.

Reply:
I have no intention of judging whether the results of these tests make sense, because that's yet another thing that people tend to disagree on and I don't want to delay tests from being added because people decide to take days or weeks to debate whether a particular result makes sense. If confirming how these abilities currently work doesn't matter, then surely #8030 doesn't need to be a stable release blocker (or be an open issue at all really).

Reply:
> If confirming how these abilities currently work doesn't matter, then surely #8030 doesn't need to be a stable release blocker (or be an open issue at all really).

So the tests are written to then update the documentation or what is the connection?

Reply:
You can't write documentation without confirming how it currently works first. And if you need to confirm how it works anyway, it might as well be done as unit tests so we know when it changes, whether it's changed intentionally or not.

Reply:
I will remove the blocker tag from that issue then.

Reply:
What are your plans going forward?

Reply:
To work on other issues with the Blocker: New Stable tag, then come back to this during 1.19.

Reply:
Regarding what Ravana said "I understand all unit tests as being about protection from unintentional changes, while not restricting anything about intentional changes." do you see it this way as well?

Do you know of more tests that check likely undesirable behavior like this case?

Would you be willing to make note if more such cases come up?

Reply:
The only tests that for sure check for such behavior are the cve test and tests that check for crashes.

Reply:
I'm very confused of what you're talking about now. I have no idea if you answered any of my questions.

1) This test checks that an attacks ability with value=0 does not reduce the strikes to 0, right?

2) This is likely undesirable behavior, right?

3) Are there more tests that you know that check for similarly undesirable behavior?

4) If you notice more such behavior would you be willing to make note of that?

5) Regarding what Ravana said "I understand all unit tests as being about protection from unintentional changes, while not restricting anything about intentional changes." do you see it this way as well?

Please answer all 5 questions.

Reply:
1. Yes.
2. I don't know. Different people think different behavior is desirable or undesirable, and since I'm not a UMC author I don't think I have any particular insight into that unless it's particularly egregious. Stuff like this or sub causing add to be ignored do seem odd to me and I might mention it that way, but I would not decide to change them or open issues to have them be changed.
3. I don't know. Same reason.
4. Yes, but I expect I would have a much higher bar than you do on judging whether a particular behavior should clearly be changed or not.
5. Yes.

Reply:
Thanks.

Mentioning odd stuff is enough because to me it is a wasted opportunity when you work out how this stuff works, notice something odd and then "sweep it under the rug" by not making any note of the oddness anywhere and even write tests like this one that hardly even protect from unintentional changes as much as protect from just one specific change that would make the behavior not odd.

Like lots of things can change for a value=0 attacks ability except if it gives the trivially expected number of 0 strikes then this test fails. It just seems like a really odd test to me. I see there are more tests with `COMPARE=numerical_not_equals` which look dubious to me as well.

Especially with no hint that this is odd behavior this seems to very easily confuse anyone in the future looking to change the behavior to something more expected (perhaps because of a bug report) and then seeing tests fail that appear specifically designed to prevent that exact change.

I think it'd be good to officially state "unit tests are about protection from unintentional changes, while not intended to restrict intentional changes" somewhere. Not sure what best to do with these tests.

Reply:
I created https://github.com/wesnoth/wesnoth/blob/1c30866f37a9031decb636dd0b77252246a93a39/data/test/scenarios/wml_tests/UnitsWML/AbilitiesWML/swarm_disables_upgrades.cfg#L18 exactly because it was odd behaviour, but I rely on it, and  hopefully be notified when this test is not valid anymore.